tests/erlang_tests/test_stacktrace.erl

```diff
@@ -0,0 +1,286 @@
+
+test_tail_recursive_throw() ->
+    ok =
+        try
+            a_tail_recursive_function(5, throw_me),
+            fail
+        catch
+            throw:throw_me:Stacktrace ->
+                expect_stacktrace(
+                    Stacktrace,
+                    [
+                        {?MODULE, maybe_crash, 1},
+                        {?MODULE, test_tail_recursive_throw, 0},
+                        {?MODULE, start, 0}
+                    ]
+                )
+        end.
+
+a_body_recursive_function(0, _Msg) ->
+    ok;
+a_body_recursive_function(I, Msg) ->
+    case a_body_recursive_function(I - 1, Msg) of
+        ok ->
+            maybe_crash(Msg);
+        _ ->
+            error
+    end.
+
+test_body_recursive_throw() ->
+    ok =
+        try
+            a_body_recursive_function(5, throw_me),
+            fail
+        catch
+            throw:throw_me:Stacktrace ->
+                expect_stacktrace(
+                    Stacktrace,
+                    [
+                        {?MODULE, maybe_crash, 1},
+                        {?MODULE, a_body_recursive_function, 2},
+                        {?MODULE, a_body_recursive_function, 2},
+                        {?MODULE, a_body_recursive_function, 2},
+                        {?MODULE, test_body_recursive_throw, 0},
+                        {?MODULE, test_body_recursive_throw, 0},
+                        {?MODULE, start, 0}
+                    ],
+                    %% TODO
+                    skip_beam
+                )
+        end.
+
+test_spawned_throw() ->
+    Self = self(),
+    spawn(
+        fun() ->
+            try
+                do_some_stuff(blah),
+                a_tail_recursive_function(5, throw_me),
+                do_some_stuff(blah),
+                Self ! fail
+            catch
+                throw:throw_me:Stacktrace ->
+                    Result = expect_stacktrace(
+                        Stacktrace,
+                        [
+                            {?MODULE, maybe_crash, 1},
+                            {?MODULE, '-test_spawned_throw/0-fun-0-', 1}
+                        ]
+                    ),
+                    Self ! Result
+            end
+        end
+    ),
+    receive
+        Result ->
+            erlang:display(Result),
+            Result
+    end.
+
+test_catch() ->
+    {'EXIT', {error_me, Stacktrace}} = (catch maybe_crash(error_me)),
+    Result = expect_stacktrace(
+        Stacktrace,
+        [
+            {?MODULE, maybe_crash, 1},
+            {?MODULE, test_catch, 0},
+            {?MODULE, start, 0}
+        ]
+    ),
+    do_some_stuff(Result),
+    Result.
+
+maybe_crash(Term) ->
+    case Term of
+        ok ->
+            ok;
+        throw_me ->
+            throw(Term);
+        error_me ->
+            error(Term)
+    end.
+
+maybe_badmatch(Term) ->
+    ok = Term.
+
+match_stacktrace([], [], is_atomvm) ->
+    ok;
+match_stacktrace(_, [], is_beam) ->
+    ok;
```

Comment:
I also noticed further cases where the stack traces are different. For example, OTP can return a first stack trace element with `{erlang, apply, Params, []}`  when apply is called, especially with non-atom module or function name.

Reply:
What I hinted here is that I see this code is very dependent of the current test runner, calling the test module directly from C for AtomVM, and invoking BEAM, and it will need to be modified if tests are run differently, something I've been toying with. That's a minor remark.

Reply:
Oh I see.  So even with AtomVM the actual stacktrace might be longer than expected, depending on the context.  Fair enough.

Reply:
I weakened the test to check instead that the created stacktrace at least contains the expected entries in the expected order.